src/lib/sql/sql_pipeline_statement.cpp

```diff
@@ -191,30 +201,67 @@ const std::shared_ptr<AbstractOperator>& SQLPipelineStatement::get_physical_plan
   return _physical_plan;
 }
 
-const std::vector<std::shared_ptr<OperatorTask>>& SQLPipelineStatement::get_tasks() {
+const std::vector<std::shared_ptr<AbstractTask>>& SQLPipelineStatement::get_tasks() {
   if (!_tasks.empty()) {
     return _tasks;
   }
 
-  _tasks = OperatorTask::make_tasks_from_operator(get_physical_plan(), _cleanup_temporaries);
+  if (_is_transaction_statement) {
+    _tasks = _get_transaction_tasks();
+  } else {
+    auto operator_tasks = OperatorTask::make_tasks_from_operator(get_physical_plan(), _cleanup_temporaries);
+    _tasks = std::vector<std::shared_ptr<AbstractTask>>(operator_tasks.cbegin(), operator_tasks.cend());
+  }
+
   return _tasks;
 }
 
+std::vector<std::shared_ptr<AbstractTask>> SQLPipelineStatement::_get_transaction_tasks() {
+  auto sql_statement = get_parsed_sql_statement();
+  const std::vector<hsql::SQLStatement*>& statements = sql_statement->getStatements();
+  auto* transaction_statement = reinterpret_cast<hsql::TransactionStatement*>(statements.front());
+  switch (transaction_statement->command) {
+    case hsql::kBeginTransaction:
+      return {std::make_shared<JobTask>([this] {
+        if (!_transaction_context || !_transaction_context->is_auto_commit())
+          this->set_warning_message(std::string("WARNING: Cannot begin transaction inside an active transaction."));
+        else
+          _transaction_context = Hyrise::get().transaction_manager.new_transaction_context(false);
+      })};
+    case hsql::kCommitTransaction:
+      return {std::make_shared<JobTask>([this] {
+        if (!_transaction_context || _transaction_context->is_auto_commit())
+          this->set_warning_message(std::string("WARNING: Cannot commit since there is no active transaction."));
```

Comment:
Why is this a warning, not an (input) assertion? Also, information should not be passed around in strings. If a string is necessary, it should be generated at the latest moment possible.

Reply:
Throwing an exception here in the regular way would cause the database to crash. This lambda is executed by the Hyrise scheduler, which has no mechanism to catch exceptions. Instead we behave in accordance with the PostgreSQL-protocol and send a special warning message packet.

I don't see a way to generate this message at a later point without introducing some additional enum class or some other kind of mapping. However this would not be consistent with regular error messags, where strings are generated at the code position where the error is thrown.

Reply:
> Throwing an exception here in the regular way would cause the database to crash. This lambda is executed by the Hyrise scheduler, which has no mechanism to catch exceptions.

Hmm, I see what you mean. If we had an AssertInput outside of the lambda, this should work, right? Can't we track `explicit_transaction_active` (naming tbd) outside of the lambda? We just need to know if there is an active transaction when we enter `get_tasks` and modify the flag accordingly when we see kCommit... whatever.

> However this would not be consistent with regular error messags, where strings are generated at the code position where the error is thrown.

Could you give me a pointer to the existing code? This doesn't sound like something we should do.

Also, one more thing: The hsql namespace should not make it outside of the SQLTranslator. Could you please translate kCommitTransaction etc. into a Hyrise-internal symbol? That improves the encapsulation.

Reply:
I tried to change the warning message into a `FailInput`, but there were a couple of things that didn't work. We couldn't send WarningMessages to the `psql` Client anymore, but that was easily fixable by adding a new `WarnInput` method. However, throwing in the sql pipeline statement breaks the pipeline and this isn't expected for a warning. Additionally, I tried to pull the check out of the lambda, but than we have nothing to return in the method and I thought adding another bool to the lambda wouldn't make it any nicer than the current solution.

If you got any other ideas, I'll gladly try to implement them.

Reply:
Maybe I have a wrong understanding of what a "warning" is in psql. My idea was to reject the execution of the current SQL string. Is the Postgres standard to just print an info message saying "Whoopsie, this didn't make sense, I just ignored your transaction statement and did potentially non-atomic auto-commit modifications to the database"? If yes, I understand what you mean. Would it be possible to deviate from that scheme and make it a hard error? I can't see any case where we would want a soft warning (that will be ignored) instead of a hard fail.

Reply:
Link to PostgreSQL error messages https://www.postgresql.org/docs/12/plpgsql-errors-and-messages.html#PLPGSQL-STATEMENTS-RAISE
What do you mean by "hard error"?
Option 1: Server crash: I think we should not do this, but be at least be a bit robust. Note, invalid sql does neither crash the server, but reports an error to the user.
Option 2: Error message to user: The error level does not change the fact that we must store the statement status.

Reply:
I'm not sure if I completely understand the issue here. If I submit an invalid SQL query through psql, I get this:

```
=> SELECT foobar;
Invalid input error: Couldn't resolve identifier 'foobar' or it is ambiguous
```

This is an AssertInput and is caught by `Session::run`. Why doesn't this work here?

Reply:
> My idea was to reject the execution of the current SQL string. Is the Postgres standard to just print an info message saying "Whoopsie, this didn't make sense, I just ignored your transaction statement and did potentially non-atomic auto-commit modifications to the database"?

So the latter is true? If you ask me, that doesn't make any sense and I can't see why a user would want that. Especially if queries are not sent manually but as part of a benchmark or an application, this seems like a recipe for desaster. In conformance with our "fail often, fail loud" principle, I would expect the database to reject such a broken query.

Reply:
> My idea was to reject the execution of the current SQL string. Is the Postgres standard to just print an info message saying "Whoopsie, this didn't make sense, I just ignored your transaction statement and did potentially non-atomic auto-commit modifications to the database"?
The current SQL statement is rejected for errors AND warnings. (This is what Postgres and we do currently.)
However, in Posgres, errors abort the current transaction. Warnings preserve the Transaction status. In both cases, the next statement is executed normally.
From the Postgres documentation "the frontend must then take responsibility for figuring out what happens if the earlier command fails and already-issued later commands succeed."

We currently abort the complete pipeline as we catch the error only in Session::run. I think it is ok for a research database. However, users often cannot directly control which statements end up in one pipeline.
The longer I think about it, I prefer using AssertInput also for warnings, although it is not the default behavior, which we also not use/guanrantee for errors.
Advantages:
- Simpler code
- The user probably did something unintentional anyway.

If you agree, I switch from warnings to errors.

Reply:
> Also, one more thing: The hsql namespace should not make it outside of the SQLTranslator. Could you please translate kCommitTransaction etc. into a Hyrise-internal symbol? That improves the encapsulation.

Note, in the current master: SQLPipeline uses hsql-things, too:
`switch (statement->type()) {
      // Check if statement alters the structure of the database in a way that following statements might depend upon.
      case hsql::StatementType::kStmtImport:
      case hsql::StatementType::kStmtCreate:
      case hsql::StatementType::kStmtDrop:
      case hsql::StatementType::kStmtAlter:
      case hsql::StatementType::kStmtRename:`
If I should change this, I would be happy about a proposal in which way to adaptd the code. Currently we use a shortcut before the statement translation happens.

Reply:
> If you agree, I switch from warnings to errors.

Agreed. To add to the advantages: This seems like the easier approach and, if we ever come to the point where it is not sufficient any more, we can still adapt it. The technical debt seems smaller than adding error handling for a case that should not exist.

> Note, in the current master: SQLPipeline uses hsql-things

Hmpf, ok. Moving this into the SQLTranslator just for the sake of it being somewhere else doesn't help anyone. Just ignore that point.

Reply:
Done.